records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,243 @@
+import json
+import time
+from contextlib import contextmanager
+from records_mover.url.base import BaseDirectoryUrl
+import logging
+from typing import Iterator, Tuple, Union, Optional, TYPE_CHECKING
+if TYPE_CHECKING:
+    from googleapiclient.discovery import _TransferJobConfig
+    import google.auth.credentials
+    from records_mover.url.gcs.gcs_directory_url import GCSDirectoryUrl
+    from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+
+logger = logging.getLogger(__name__)
+
+
+class CopyOptimizer:
+    """When Records Mover needs to move data between different URL
+    schemes, the default way (a single-threaded downloading of bytes
+    from the source and then reuploading of bytes to the target) can
+    be very slow for large directories.
+
+    This class can use faster means (ideally a cloud-based service
+    that can copy things at scale, or at least vendor-optimized
+    multi-threaded upload/download code) when they are available.
+
+    This depends on the specifics of the copy, and as such return
+    values must be checked to determine if the copy was able to be
+    performed.
+    """
+    def copy(self,
+             loc: BaseDirectoryUrl,
+             other_loc: BaseDirectoryUrl) -> bool:
+        # TODO: document
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker